06-wrappers/06-UiInputDate/components/UiInputDate.vue

```diff
@@ -1,13 +1,101 @@
 <template>
-  <UiInput />
+  <UiInput v-model="computedModelValue" :type="type" :step="stepValue" ref="input">
+    <template v-if="$slots['left-icon']" #left-icon>
+      <slot name="left-icon" />
+    </template>
+    <template v-if="$slots['right-icon']" #right-icon>
+      <slot name="right-icon" />
+    </template>
+  </UiInput>
 </template>
 
-<script>
+<script lang="ts">
+import dayjs from 'dayjs';
+import utc from 'dayjs/plugin/utc';
+import { defineComponent } from 'vue';
 import UiInput from './UiInput.vue';
 
-export default {
+dayjs.extend(utc);
+
+const DateKind = {
+  DATE: 'date',
+  TIME: 'time',
+  DATETIME_LOCAL: 'datetime-local',
+} as const;
+
+export default defineComponent({
   name: 'UiInputDate',
+  props: {
+    modelValue: Number,
+    type: {
+      type: String,
+      default: DateKind.DATE,
+      validate(value: string): boolean {
+        return value in DateKind;
+      },
+    },
+    step: {
+      type: Number,
+      default: 1,
+    },
+  },
+
+  emits: ['update:modelValue'],
 
   components: { UiInput },
-};
+
+  computed: {
+    dateValue(): number {
+      return dayjs(this.modelValue).utc().startOf('day').valueOf();
+    },
+
+    timeValue(): number {
+      return dayjs(this.modelValue).utc().valueOf() - this.dateValue;
+    },
+
+    secondValue(): number {
+      return dayjs(this.modelValue).utc().diff(dayjs(this.modelValue).utc().startOf('minute'), 'milliseconds');
+    },
+
+    stepValue(): number {
+      return this.type === DateKind.TIME ? 60 * this.step : this.step;
+    },
+
+    computedModelValue: {
+      get(): string {
+        if (!this.modelValue) {
+          return '';
+        }
+
+        switch (this.type) {
+          case DateKind.TIME:
+            return dayjs(this.modelValue).utc().format('HH:mm');
+
+          case DateKind.DATETIME_LOCAL:
+            return dayjs(this.modelValue).utc().format('YYYY-MM-DDTHH:mm:ss');
+
+          default:
+          case DateKind.DATE:
+            return dayjs(this.modelValue).utc().format('YYYY-MM-DD');
+        }
+      },
+
+      set() {
+        let inputValue: number = ((this.$refs.input as typeof UiInput).$refs.input as HTMLInputElement).valueAsNumber;
+
+        switch (this.type) {
+          case DateKind.TIME:
+            inputValue += this.dateValue + this.secondValue;
+            break;
+
+          case DateKind.DATE:
+            inputValue += this.timeValue;
+            break;
+        }
```

Comment:
I suppose, these line do nothing, because `dateValue` is supposed to be 0 for `input[type=time]` and `secondValue` should be 0 if step integer.

By the way, with float step milliseconds are already included in the `valueAsNumber`...

Reply:
I noticed that `valueAsNumber` contains value reduced depending on type (only time or only date). So we need to correct the value to full date value.

If I change the date in the first input - the time in the second input will be reset:
![image](https://user-images.githubusercontent.com/47382827/231761128-f0d685dc-f505-4770-a047-854aee125160.png)

Reply:
> So we need to correct the value to full date value.

According to the specification, it is not supposed to be a full date.

When it is a time picker, the date is always null (1970-01-01).

When it is a date ticker, the time is always null (00:00:00.000).

And only datetime picker has both (depends on the step, it may have null seconds and null ms).

The idea is to guarantee that to selected times (12:00 12:00) are always equals (and it is not actually something like `1970-01-01T12:00` and `2020-02-02T:12:00`).

Reply:
If I select some HH:MM time in one datepicker and select the same time in another time picker, I should have exactly the same value (same UNIX timestamp)

Reply:
Specs: https://html.spec.whatwg.org/multipage/input.html#time-state-(type=time)

> The algorithm to convert a string to a number, given a string input, is as follows: ... ... **the number of milliseconds elapsed from midnight to the parsed time on a day with no time changes.**

So the value cannot be more than 24 * 60 * 60 * 1000 ms

Reply:
I get it! I thought `modelValue` should always contain full date value.